ent/encryption/encrypted_file_impl.cc

```diff
@@ -701,5 +701,82 @@ std::unique_ptr<data_sink_impl> make_encrypted_sink(data_sink sink, shared_ptr<s
     return std::make_unique<encrypted_data_sink>(std::move(sink), std::move(k));
 }
 
+class encrypted_data_source : public data_source_impl, public block_encryption_base {
+    data_source _source;
+    std::vector<char> _buffer;
+    size_t current_position = 0;
+
+public:
+    encrypted_data_source(data_source source, shared_ptr<symmetric_key> k) : block_encryption_base(std::move(k)), _source(std::move(source)) {
+        _buffer.reserve(block_size * 2);
+    }
+
+    future<temporary_buffer<char>> get() override {
+        while (_buffer.size() < block_size) {
+            auto buf = co_await _source.get();
+            if (buf.empty()) {
+                break;
```

Comment:
You need to keep track of this. If this is EOS, you need to special case it below.